### ISSUE TITLE
SSL error: error:140E0114:SSL routines:SSL_shutdown:uninitialized

### ISSUE THREAD

**Issue by AaronO** [libgit2/git2go#252]
Fetching a HTTPS repo fails with the current `master` and `libgit2`@`6b36945d7e8a780a3063f09c65b3d9fed6525552`

```
SSL error: error:140E0114:SSL routines:SSL_shutdown:uninitialized
```

Any idea of the source of this issue ? Would this be fixed by https://github.com/libgit2/git2go/pull/152 ?

**Comment by carlosmn** [comment]
That PR shouldn't change what we do, just how we do it. You can try it out yourself by including an OpenSSL-initialising package in your program.

**Comment by AaronO** [comment]
Any idea where the source of the issue could be then ?

**Comment by carlosmn** [comment]
This can apparently come up when there is no network connectivity. It could be that we're trying to deallocate an `SSL` object which was not allocated yet, which would indicate that this might be a bug in our connection code. Or it could be something completely different.

Is this happening repeatedly? Is the host you're trying to connect up? Accepting connection?

**Comment by AaronO** [comment]
After debugging further, it does seem to happen when there's no network or a DNS issue in the docker container (the same thing).

But I actually encountered this bug before when statically linking to `libssl` on linux.

The dynamically linked binary would work perfectly but the static one would fail, not immediately but on the second call to `Fetch`.

So maybe there's a deeper issue. (I can deal with the networking one, static linking isn't a must but would be nice, so nothing too urgent). (I'll take a dive into `libgit2` later next week).

**Comment by carlosmn** [comment]
That's a strong indicator that we end up calling `SSL_shutdown()` due to cleanup for failed connections, so we end up hiding the real error.

**Comment by carlosmn** [comment]
This ended up being that OpenSSL doesn't like us calling `SSL_shutdown()` on a `NULL` context when we fail to look up a hostname or similar. This is fixed in libgit2/libgit2#3445 which is included in the vendored libgit2 in `next` and in libgit2's maintenance releases.
